text/0024-track-validator-set-when-synchronizing.org

```diff
@@ -0,0 +1,141 @@
+
+#+BEGIN_SRC svgbob :file images/0024/joining.svg :exports results
+
+            Validators    Blocks       Joining Node Actions
+
+       __
+       \/   .~~~~~~~.     ,-------.
+       __   : A B C :~~~~/ Era 1 /<--- Initialize validator set
+       \/   `~~~~~~~'   '----+--'
+       __                    |
+ -+-   \/                   [#]<------ Check A/B/C sigs
+  |    __                    |
+       \/                   [#]<------ Check A/B/C sigs
+  |    __                    |
+  |    \/     .~~~~~.     ,--+----.
+       __     : A B :~~~~/ Era 2 /<--- Update validator set
+ |\/|  \/     `~~~~~'   '----+--'
+ |  |  __                    |
+       \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+ |--   \/                   [#]<------ Check A/B sigs
+ +--   __                    |
+       \/   .~~~~~~~.     ,--+----.
+       __   : A B D :~~~~/ Era 3 /<--- Update validator set
+       \/   `~~~~~~~'   '----+--'
+       __                    |
+       \/                   [#]<------ Check A/B/D sigs
+       __                    :
+       \/                    :
+
+
+  +--------+-------------------------------------+
+  | Legend |                                     |
+  +--------+                                     |
+  |                                              |
+  |                          |                   |
+  |  .~~~~.                 ,+--.                |
+  |  :    : Validator Set  /   /   Switch Block  |
+  |  `~~~~'               '--+'                  |
+  |                          |                   |
+  |                                              |
+  |                          |                   |
+  |                         [#]    Normal Block  |
+  |                          |                   |
+  |                                              |
+  +----------------------------------------------+
+
+
+#+END_SRC
+
+#+RESULTS:
+[[file:images/0024/joining.svg]]
+
+As we can see above, switch block can change the validator
+set. Joining nodes needs to track this.
+
+** Reference-level explanation
+:PROPERTIES:
+:CUSTOM_ID: reference-level-explanation
+:END:
+
+/Synchronization mode/ is mediated by the variant
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/node/src/components/linear_chain_sync.rs#L85-L93][=linear_chain_sync::State::SyncingDescendents]]=.
+
+We could add a field =validators= to this with type
+[[https://github.com/xcthulhu/casper-node/blob/0a7f9e5fd7608e2f6574c1e213bd9f5e35880af5/types/src/auction/types.rs#L15][=types::ValidatorWeights=]] (i.e. =BTreeMap<PublicKey, U512>=). This would
+be updated as switch blocks are processed.
+
+To support this, blocks will include the list of validators whenever
+they are updated.  There are two scenarios where this can happen:
+
+  1. In a switch block when new validators become bonded/unbonded
+  2. When a validator has been slashed, although this may not be necessary
```

Comment:
@afck my understanding of the protocol is that this is not necessary...

Reply:
Yes, I'd actually remove point 2: From the protocol's perspective, slashing does _not_ immediately update the validator set.
(The protocol internally ignores the faulty validator's votes, but that's different.)

Reply:
Clarified in https://github.com/CasperLabs/ceps/pull/24/commits/9a2daaeb1719b995c9008c82d6045f4eef2cc6b4